### DISCUSSION TITLE
How to check sensor (DS18B20) availability?  - SCRIPT

### DISCUSSION THREAD

**Discussion by mmmmm009** [arendst/Tasmota#20892] [Q&A]
Hi,

I have a DS18B20 sensor connected to my ESP32.
I read it in T section of my SCRIPT

`>T
temperature=DS18B20#Temperature`

How can I check if sensor is available at first?
(to prevent running the rest of my script in case of loosing the sensor)

**Comment by HofeBY** [comment]
Did you add a resistor of 4,7kOhm betwen 3,3V (red) and Signal Pin (yellow)?
GPIO Pin is configured as D18Bx20 1 and Signal line is connected correct ?
If your Tasmota Version is compiled with DS18B20 Suppport it schould been seen on main screen and in Tasmota Log.

If all is functional, you can read in Temperature values in section >T

For used Name take a look in Tasmota Log.

**Comment by mmmmm009** [comment]
Sensor is connected and running well.
However I'd like to create some kind of "sensor fail alert"

So before I get the reading from sensor I'd like to chceck if sensor is still present.

**Reply by sfromis** [reply]
In general, the better method for detecting an outage is when data is not arriving within the expected time frame, like a timer being reset when getting data, and expiry of the timer means dead.

**Comment by mmmmm009** [comment]
```
def ReadTemp ()
var sensors = json.load(tasmota.read_sensors())  
temperature = (sensors['DS18B20']['Temperature'])
print (temperature)
end
```
When I disconnect gnd (simulating sensor failure) I get:
Berry console:
```
ReadTemp()
BRY: Exception> 'key_error' - DS18B20
stack traceback:
	<native>: in native function
	<unknown source>: in function `ReadTemp`
	<unknown source>: in function `main`
```	
What should the condtion to check sensor/temp be like?

```
def ReadTemp ()
var sensors = json.load(tasmota.read_sensors())  

IS SENSOR VALID?

temperature = (sensors['DS18B20']['Temperature'])
print (temperature)
end
```

**Reply by sfromis** [reply]
What happens with the Berry coding is that when you are not getting values from the sensor, the map from `json.load` will not have such data. One technique I like to use to safeguard against any way of getting `key_error` is coding like this:
`temperature = sensors.find('DS18B20',{}).find('Temperature',nil)`
By providing a default value, I can get this instead of a failure. Arguably, getting a `DS18B20` object not having a `Temperature` value should not be possible, but I preferably only use `[]` for map references when I've put the data into it myself....

Another approach could be based on `if sensors.contains('DS18B20')` to check.

**Comment by mmmmm009** [comment]
```def ReadTemp ()
var sensors = json.load(tasmota.read_sensors())  

if sensors.contains('DS18B20')
    temperature = (sensors['DS18B20']['Temperature'])
    print ("Sensor OK")
else
   print ("Sensor missing")
end

end
```

thx.
